File: design/src/lib/runninggu/poi.js
// POI 어댑터 — 폴백 체인: ① 카카오맵 실시간 → ② 사전수집(raceId별) → ③ 합성 샘플.
// 어떤 소스인지 source('live'|'sample'|'synth')로 반환해 결과 화면에서 배지 표기.
import { PRESAMPLED, synthPOIs } from './sampleData.js'

// autoload=false 로 로드된 카카오 SDK 준비 대기.
let kakaoReady = null
export function ensureKakao(timeout = 4000) {
  if (kakaoReady) return kakaoReady
  kakaoReady = new Promise((resolve) => {
    if (typeof window === 'undefined') return resolve(null) // SSR/Node: 폴백 사용
    const start = Date.now()
    const tick = () => {
      const k = window.kakao
      if (k && k.maps && k.maps.services) return resolve(k)
      if (k && k.maps && k.maps.load) {
        k.maps.load(() => resolve(window.kakao))
        return
      }
      if (Date.now() - start > timeout) return resolve(null)
      setTimeout(tick, 120)
    }
    tick()
  })
  return kakaoReady
}

function fromKakao(r) {
  return {
    name: r.place_name,
    lat: Number(r.y),
    lng: Number(r.x),
    desc: (r.category_name || '').split('>').pop()?.trim() || '',
    addr: r.road_address_name || r.address_name || '',
    url: r.place_url || '',
  }
}

// 카카오 실시간 검색 (카테고리 code 우선, 없으면 kw 키워드).
function kakaoSearch(kakao, cat, center, count) {
  return new Promise((resolve, reject) => {
    const ps = new kakao.maps.services.Places()
    const loc = new kakao.maps.LatLng(center.lat, center.lng)
    const opts = { location: loc, radius: 8000, size: Math.min(count, 15), sort: kakao.maps.services.SortBy.DISTANCE }
    const cb = (data, status) => {
      if (status === kakao.maps.services.Status.OK) resolve(data.map(fromKakao))
      else reject(new Error('kakao:' + status))
    }
    if (cat.code) ps.categorySearch(cat.code, cb, opts)
    else ps.keywordSearch(cat.kw || cat.label, cb, opts)
  })
}

// 단일 카테고리 조회. { source, places } 반환.
export async function searchPOIs({ cat, center, raceId, count = 8 }) {
  const kakao = await ensureKakao()
  if (kakao) {
    try {
      const live = await kakaoSearch(kakao, cat, center, count)
      if (live && live.length) return { source: 'live', places: live.slice(0, count) }
    } catch {
      /* 폴백으로 진행 */
    }
  }
  const pre = PRESAMPLED[raceId] && PRESAMPLED[raceId][cat.key]
  if (pre && pre.length) return { source: 'sample', places: pre.slice(0, count) }
  return { source: 'synth', places: synthPOIs(cat.key, center, count) }
}

File: android/app/src/main/java/com/runninggu/app/domain/Poi.kt
package com.runninggu.app.domain

/** 위경도. 내부 표준은 `lat`/`lng` 이고 변환은 경계에서만 한다. (SPEC §6.6) */
data class LatLng(val lat: Double, val lng: Double)

/**
 * 장소 하나. (SPEC §6.3 `Poi`)
 *
 * @param desc 카카오 `category_name` 말단 또는 KTO 분류. 블록 설명의 기본값으로 쓴다.
 */
data class Poi(
    val name: String,
    val lat: Double,
    val lng: Double,
    val desc: String = "",
    val addr: String = "",
    val url: String = "",
)

/**
 * POI 를 어디서 얻었는지. 결과 화면이 배지로 노출한다. (SPEC §6.3 · NFR-2)
 *
 * 원본 웹 구현은 소문자(`live`·`sample`·`synth`)를 쓰지만 §6.3 이 **Enum 대문자 통일**을
 * 요구하므로 포팅하며 맞춘다.
 */
enum class PoiSourceKind { LIVE, SAMPLE, SYNTH }

/** 카테고리 한 건의 조회 결과. (SPEC §6.3) */
data class PoiResult(
    val source: PoiSourceKind,
    val places: List<Poi>,
)

/**
 * 동선 엔진이 쓰는 POI 공급원.
 *
 * **엔진이 직접 조회하지 않는 이유** — 원본 `engine.js` 는 함수 안에서 네트워크를 부른다.
 * 그대로 옮기면 `domain` 이 네트워크에 묶여 단위 테스트가 불가능해진다. 조회를 밖으로 빼고
 * 엔진은 받은 POI 로 조립만 한다.
 *
 * 실제 구현은 `data/remote` 가 서버를 거쳐 채운다(AP-14). 앱은 카카오·KTO 를 직접 부르지 않는다.
 * 테스트에서는 가짜 구현을 끼운다.
 */
interface PoiSource {
    suspend fun load(category: PoiCategory, center: LatLng, count: Int): PoiResult
}
